### PR TITLE
Throw a logic error when calling Rewind on unsupported BodyStreams and update the exception message to be actionable.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/Azure/azure-sdk-for-cpp/issues/1704
Throwing `invalid_argument`, as the issue described, didn't fit well, especially since the API doesn't accept any arguments directly.

Other alternatives are just throw an `std::exception` (base type of `runtime_error`).

### CODE REVIEW

**Review by vhvb1989 — APPROVED** [review]
Add changelog and optionally a test case

**Comment by vhvb1989** [comment]
@JeffreyRichter would it be bad to upgrade our curl and winhttp body streams to support rewind? It already has the connection handle, it only need to keep a copy of the request and re send.. supposing we don't consider any headers like date or things like that important to be maintained

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp [hunk 1]
```diff
@@ -68,10 +68,10 @@ namespace Azure { namespace Core { namespace Http {
      */
     virtual void Rewind()
     {
-      // Exception is not meant to be caught. This is rather an indication of a bad program that
-      // needs to be updated. This is the reason of using `throw` alone.
-      throw "The upload BodyStream doesn't support Rewind which is required to guarantee fault "
-            "tolerance when retrying the operation.";
+      throw std::runtime_error(
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:71]: std::invalid_argument makes more sense.  Its a logic error so should make sense for folks to handle that at dev time

**Review by RickWinter — APPROVED** [review]
Switch to invalid_argument before checking in.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp (hunk 1 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:71]: Imagine the user calls `stream.Rewind()` and get's back `invalid_argument`. How would they reason about that? They didn't pass in any arguments.

It feels strange to me we'd throw such an exception, even if it is something the caller needs to handle at dev time.

**Comment by ahsonkhan** [comment]
> Add changelog 

I thought it and his isn't too relevant from an end customer perspective, particularly since it is just a change in exception/message.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp (hunk 1 above)
Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:71]: runtime_error exceptions are for
```
std::runtime_error
Defines a type of object to be thrown as exception. It reports errors that are due to events beyond the scope of the program and can not be easily predicted.
```

In this particular case the user did something in their code that can be predicted and thus logic_error is the type of exception we should use
```
std::logic_error
Defines a type of object to be thrown as exception. It reports errors that are a consequence of faulty logic within the program such as violating logical preconditions or class invariants and may be preventable.
```

**Pull request merged**